_posts/2017-10-04-scalamacros.md

```diff
@@ -0,0 +1,299 @@
+into a single case class
+<blockquote class="twitter-tweet" data-lang="en"><p lang="en" dir="ltr">New style &quot;inline/meta&quot; macro merged into scalafmt, Dotty-ready macros are happening and they are amazing! <a href="https://twitter.com/hashtag/scala?src=hash">#scala</a> <a href="https://github.com/scalameta/scalafmt/pull/459">scalameta/scalafmt#459</a> <a href="https://t.co/ZvjdWE0Gp3">pic.twitter.com/ZvjdWE0Gp3</a></p>&mdash; Ólafur Páll Geirsson (@olafurpg) <a href="https://twitter.com/olafurpg/status/779372897198637057">September 23, 2016</a></blockquote>
+
+Many other OSS Scala libraries also use Scalameta macro annotations to
+to maximize type safety, expressiveness and performance:
+- [Freestyle](http://frees.io/):  cohesive & pragmatic framework of FP centric
+  Scala libraries
+- [Stalagmite](https://vovapolu.github.io/scala/stalagmite/perf/2017/09/02/stalagmite-performance.html):
+  effective and customizable replacement of conventional case classes with
+  convenient optimizations.
+- [Mainecoon](http://kailuowang.com/mainecoon/): A library for transforming and
+  composing tagless final algebras
+- [Dilate](https://github.com/vitorsvieira/dilate) /
+  [Newtypes](https://github.com/alexknvl/newtypes): better value
+  classes/newtypes for Scala.
+- [Example.scala](https://static.javadoc.io/com.thoughtworks.example/unidoc_2.12/2.0.0/com/thoughtworks/example.html):
+  Generate unit tests from Scaladoc strings
+
+The Scala community is indeed creative and eager to explore new metaprogramming
+facilities.
+
+Scalameta macro annotations are far from perfect, they suffer from integration
+problems with Scaladoc, Scala IDE/presentation compiler, 
+Scala REPL and other compiler plugins such as Scoverage.
+Most importantly however, we discovered several drawbacks with the converter-based
+approach while adding support for def macros with access to the semantic API.
+The issues we encountered are documented in more detail in section 3 of the
+technical report ["Two Approaches to Portable Macros"].
+
+Nevertheless, we learned a few important lessons
+
+- Users and macro authors care a lot about IDE support, in particular IntelliJ
+  support.
+- Building a portable macro system with ASTs that don't leak compiler
+  implementation details is possible.
+- Eagerly converting between the compiler and scalameta ASTs introduces
+  a lot of problems that can be avoided by implementing macros in terms of
+  extractor/constructor methods on *abstract syntax*.
+
+With these lessons learned, we decided to retire our efforts to build a macro
+system on top of Scalameta.
+From now on, Scalameta's primary focus is to support building developer tools
+such as Scalafmt, Scalafix and Metadoc.
+
+### v3: scala.macros
+
+In winter and spring 2017, Eugene Burmako at Twitter and Liu Fengyun at EPFL
+worked on a new macro system to address the limitations of scala.reflect-based
+and scala.meta-based macros.
```

Comment:
For clarity, I would say:

"worked on a new macro system to address the limitations of scala.reflect-based (v1) and scala.meta-based (v2) macros."

Reply:
Done.